sql: ensure sub-plans of CREATE TABLE AS and CREATE VIEW are always closed

knz opened the pull request:
Addresses an issue found while investigating #10466:

https://github.com/cockroachdb/cockroach/issues/10466#issuecomment-258868483

<!-- Reviewable:start -->

---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/10527)
<!-- Reviewable:end -->

knz committed:
sql: ensure sub-plans of CREATE TABLE AS and CREATE VIEW are always closed.

a-robinson commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> other than one question

---

Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/sql/create.go, line 634 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10527#-KW1MikT8S9_hOsZGksD:-KW1MikUdbRFIGS1FQm2:b-ohx5m4) ([raw file](https://github.com/cockroachdb/cockroach/blob/280706a57fe43d67e742c844582cd048232b595a/pkg/sql/create.go#L634)):*
> ```Go
> 		n.sourcePlan = nil
> 
> 		desiredTypesFromSelect := make([]parser.Type, len(resultColumns))
> ```

Is removing this logic related? Why is it not needed anymore? It might deserve a callout in the commit message (or a separate commit if it is indeed unrelated)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10527#-:-KW1N8F3lV8QkfAG8AQy:b1wqfxc)*
<!-- Sent from Reviewable.io -->

knz committed:
sql: fix a minor semantic issue in CREATE TABLE AS.

Prior to this patch, the code for CREATE AS would give the column
signature of the source table as "desired type" for the internal
insertNode it uses to populate the new table. This is conceptually
problematic, as the source table is the source to the insertNode, and
not its consumer. It is also entirely ineffectual in this case,
because the internal insertNode used by CREATE TABLE AS does not use a
RETURNING clause and thus cannot used the "desired types" provided by
its context.

This patch removes this connection.

knz commented:
TFYR!

---

Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/sql/create.go, line 634 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10527#-KW1MikT8S9_hOsZGksD:-KW2QB21ubM1xafQEukq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/280706a57fe43d67e742c844582cd048232b595a/pkg/sql/create.go#L634)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary>
> Is removing this logic related? Why is it not needed anymore? It might deserve a callout in the commit message (or a separate commit if it is indeed unrelated)
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10527#-:-KW2QSQqAtY3D2z8pNzQ:b-814cj3)*
<!-- Sent from Reviewable.io -->
